bots/tests-scan

```diff
@@ -358,22 +339,69 @@ def cockpit_tasks(api, update, branch_contexts, repo, pull_data, pull_number, sh
 
             return False
 
-        if trigger_externals():
+        def get_externals():
+            result = []
             for proj_repo in testmap.projects():
                 if proj_repo == "cockpit-project/cockpit":
                     continue
                 for context in testmap.tests_for_project(proj_repo).get("master", []):
-                    repo_context = context + "@" + proj_repo
-
-                    status = statuses.get(repo_context, {})
-                    if login not in whitelist and status.get("description", github.NO_TESTING) == github.NO_TESTING:
-                        priority = github.NO_TESTING
-                        changes = { "description": github.NO_TESTING, "context": repo_context, "state": "pending" }
-                    else:
-                        (priority, changes) = prioritize(status, title, labels, baseline, repo_context, number)
-                    if not update or update_status(api, revision, repo_context, status, changes):
-                        # don't specify a ref to check out; make-source already checks out project master
-                        results.append((priority, "pull-%d" % number, number, revision, "", context, base, proj_repo, "pull/%s/head" % number))
+                    result.append(context + "@" + proj_repo)
+            return result
+
+        def is_valid_context(context):
+            return context.split("@")[0] in contexts
+
+        # Create list of statuses to process
+        todos = {}
+        for status in statuses: # Firstly add all valid contexts that already exist in github
+            if is_valid_context(status):
+                todos[status] = statuses[status]
+        if not statuses: # If none defined in github add basic set of contexts
+            for context in branch_contexts.get(base, contexts):
+                todos[context] = {}
+        if trigger_externals():
+            for context in get_externals():
+                if context not in todos:
+                    todos[context] = {}
+
+        for context in todos:
+            # Get correct project and branch. Ones from test name have priority
+            project = repo
+            branch = base
+            (os_scenario, _, override) = context.partition("@")
+            override = override.split("/")
+            if len(override) == 2:
+                project = "/".join(override)
+            elif len(override) == 3:
+                project = "/".join(override[:2])
+                branch = override[2]
+
+            # We need to distinguish between external projects or stable branches,
+            # since those need bots_ref and not ref
+            ref = "pull/%d/head" % number
+            is_stable = base != branch
+            needs_bots = repo != "cockpit-project/cockpit" or is_stable
```

Comment:
The *only* case where a bots/ checkout isn't needed is a PR on cockpit-project/cockpit/master. As soon as we split out bots/ to a separate project, this special case will go away. I still don't understand the purpose of "is_stable", but it really feels like bots/ should not be connected to this concept.

It even feels like the whole "check out a particular bots ref" could be done for cockpit/master PRs even -- it would basically overwrite bots/ with an identical copy, but that doesn't really hurt and would make the whole thing simpler?